src/bilinearinterpolationframe.cpp

```
@@ -0,0 +1,288 @@
+                           1);
+  const int64_t down_sample_x =
+      static_cast<int64_t>(ceil(static_cast<double>(levelWidth_) /
+                                static_cast<double>(targetLevelWidth_)) -
+                           1);
+
+  // Extend samping with to include boundry voxels.  Sample sampleWidth &
+  // sampleHeight = target sampeling region for level to downsample
+  int64_t sampleWidth = 2 * down_sample_x + frameWidthDownsampled_;
+  int64_t sampleHeight = 2 * down_sample_y + frameHeightDownsampled_;
+
+  // Adjust sampleing region and start position to include surrounding voxels
+  int64_t Level_x = locationX_ - down_sample_x;
+  int64_t Level_y = locationY_ - down_sample_y;
+  if (Level_x < 0) {
+    sampleWidth += Level_x;
+    Level_x = 0;
+  }
+  if (Level_y < 0) {
+    sampleHeight += Level_y;
+    Level_y = 0;
+  }
+  if (Level_x + sampleWidth >= levelWidth_) {
+    sampleWidth = levelWidth_ - Level_x - 1;
+  }
+  if (Level_y + sampleHeight >= levelHeight_) {
+    sampleHeight = levelHeight_ - Level_y - 1;
+  }
+
+  // Allocate memory to retrieve layer data from openslide
+  uint32_t *buf_bytes = reinterpret_cast<uint32_t *>(
+      malloc((size_t)sampleWidth * (size_t)sampleHeight * sizeof(uint32_t)));
+
+  // Open slide API samples using xy coordinages from level 0 image.
+  // upsample coordinates to level 0 to compute sampleing site.
+
+  const int64_t Level0_x = (Level_x * level0Width_) / levelWidth_;
+  const int64_t Level0_y = (Level_y * level0Height_) / levelHeight_;
+  // Open slide read region returns ARGB formated pixels
+  // Values are pre-multiplied with alpha
+  // https://github.com/openslide/openslide/wiki/PremultipliedARGB
+  openslide_read_region(osr_, reinterpret_cast<uint32_t *>(buf_bytes), Level0_x,
+                        Level0_y, level_, sampleWidth, sampleHeight);
+
+  if (openslide_get_error(osr_)) {
+    BOOST_LOG_TRIVIAL(error) << openslide_get_error(osr_);
+  }
```

should we throw an exception here?

---

sure added a throw